### ISSUE TITLE
Using React Hooks in React Native

### ISSUE THREAD

**Issue by theluk** [facebook/react-native#21967]
- [x] Review the documentation: https://facebook.github.io/react-native
  - [x] Search for existing issues: https://github.com/facebook/react-native/issues
  - [x] Use the latest React Native release: https://github.com/facebook/react-native/releases

## Environment
  [skip envinfo]
  OS: Windows 10
  Node: 10.11.0
  Yarn: Not Found
  npm: 6.4.1
  Watchman: Not Found
  Xcode: N/A
  Android Studio: Version  3.2.0.0 AI-181.5540.7.32.5056338

Packages: (wanted => installed)
  react: 16.7.0-alpha.0 => 16.7.0-alpha.0
  react-native: https://github.com/expo/react-native/archive/sdk-30.0.0.tar.gz => 0.55.4
  expo: 2.2.0

## Description
When using React Hooks like `useState` within a function component, I get the following error message.

> hooks can only be called inside the body of a function component

```
export default function RootContainer(props) {
  const [something, setSomething] = useState('nothing');

  return (
    <View />
  );
};
```

**Comment by brunolemos** [comment]
> hooks can only be called inside the body of a function component

I noticed that this error also happens when you have two different instances of React being used, to double check that. (https://github.com/facebook/react/issues/13991)

I was not able to make it work on react-native too, but with a different error:

`TypeError: dispatcher.useState is not a function`

<img height="500" src="https://user-images.githubusercontent.com/619186/47597428-fed30780-d964-11e8-84a9-b60c874063ae.png" />

**Comment by developer239** [comment]
😢😭

**Comment by hramos** [comment]
We need to update the Renderer alongside React. Please wait for the React sync to land, and we will cut a new official release with 16.6 then.

**Comment by brunolemos** [comment]
> We’ll do the same once 16.7 stable is released.

It would be much appreciated to have a `next` version of react-native synced with the `next` version of `16.7` so we can try hooks right now.

My project is universal (same components between web and mobile) so my hands are tied. Can't use hooks because react-native doesn't support it.

**Comment by gaearon** [comment]
We can just do a sync with `enableHooks: false` feature flag and then those who want can override it locally.

**Comment by ferrannp** [comment]
> We’ll do the same once 16.7 stable is released.

@hramos React team ask to the community to try hooks and give feedback about them. IMO we need a way for RN people to be able to provide this feedback too (not only React projects) before anything is stable.

**Comment by fungilation** [comment]
Hooks is quite a change for React / RN syntax. A lot of docs will need updating too?

**Comment by hramos** [comment]
If you'd like to try out Hooks today, you can do the following after installing React 16.7.0-alpha in your project:

1. Clone the https://github.com/facebook/react repository. 
2. Navigate to the `react` directory.
3. Run `yarn install`
4. Run `yarn build -- --type=RN_OSS`. 
5. After a few seconds, you'll notice a new `build/react-native/` directory has been created. Copy the contents of this directory, to your project's `node_modules/react-native/Libraries/Renderer` directory.

Let us know if the above steps work out for you.

---

We're planning on landing support for React 16.6.0 in 0.57.5. Hooks is available in the 16.7.0-alpha, and feedback from the community so far has been that upgrading to alphas of React in the stable React Native releases can be disruptive. What do people here think about a React Native 0.57.5-alpha release that includes support for 16.7.0-alpha out of the box?

**Comment by dabit3** [comment]
Thanks @hramos for this info! So this works fine when using `useState`, but for some reason when I move on to try `useEffect` I get the following error:

```
failed to execute 'postMessage' on 'DedicatedWorkerGlobalScope': No function was found that matched the signature provided.
```

Any ideas on what could be causing this or how to mitigate?

**Comment by jlongster** [comment]
Wouldn't it be something like a 0.57.6-alpha if you're already aiming for a stable 0.57.5?

I'm totally fine using a custom build for React Native, as long as it works. Haven't tried it yet but does the error in the above comment make sense? If it helps I can try it as well.

**Comment by Manishalexin** [comment]
> Thanks @hramos for this info! So this works fine when using `useState`, but for some reason when I move on to try `useEffect` I get the following error:
> 
> ```
> failed to execute 'postMessage' on 'DedicatedWorkerGlobalScope': No function was found that matched the signature provided.
> ```
> Any ideas on what could be causing this or how to mitigate?

I was able to use useEffect without any issues. Anything in particular that didn't work with useEffect?

**Comment by dabit3** [comment]
@Manishalexin Just tried it again with a completely new build, still not working for me.

I've tried useEffect in a couple of different ways, still getting errors on all of them (same error as above).

```
React Native Environment Info:
    System:
      OS: macOS Sierra 10.12.6
      CPU: x64 Intel(R) Core(TM) i5-7360U CPU @ 2.30GHz
      Memory: 52.84 MB / 8.00 GB
      Shell: 5.2 - /bin/zsh
    Binaries:
      Node: 8.12.0 - ~/.nvm/versions/node/v8.12.0/bin/node
      Yarn: 1.3.2 - /usr/local/bin/yarn
      npm: 6.4.1 - ~/.nvm/versions/node/v8.12.0/bin/npm
    SDKs:
      iOS SDK:
        Platforms: iOS 11.2, macOS 10.13, tvOS 11.2, watchOS 4.2
      Android SDK:
        Build Tools: 23.0.1, 26.0.1, 26.0.2, 26.0.3, 27.0.3
        API Levels: 23, 26, 27
    IDEs:
      Android Studio: 3.0 AI-171.4443003
      Xcode: 9.2/9C40b - /usr/bin/xcodebuild
    npmPackages:
      react: ^16.7.0-alpha.0 => 16.7.0-alpha.0
      react-native: 0.57.4 => 0.57.4
    npmGlobalPackages:
      react-native-cli: 2.0.1
```

**Comment by Manishalexin** [comment]
@dabit3 I have the exact same config for npm modules but system-wise I am Mojave and Xcode 12( Shouldn't matter, I presume). And moreover, the error seems service worker related. ( which is weird in vanilla RN project )
I'll quickly try with a fresh project and report back. ( I tried it with an existing project )

**Comment by dabit3** [comment]
@Manishalexin did you see my comment about the remote debugger?

**Comment by Manishalexin** [comment]
@dabit3 Just saw. ( Comment doesn't update till page refresh ).

**Comment by IljaDaderko** [comment]
@hramos Regarding your question of using react alpha in react-native: I am usually freakishly strict about using unstable versions of packages in my projects, but having played with 16.7 alpha this weekend I didn't notice any bugs.

Perhaps an alpha release of react-native including hooks could be a good idea? I assume if hooks go past RFC stage and get implemented, react-native will need a lot of pre-testing and this seems like a feature that people would actually love to start using and testing as early as now?

**Comment by jfrolich** [comment]
An alpha release of react-native would be awesome! I am working on a new app that isn't published yet, and would love to use the new hooks API.

**Comment by chrisbianca** [comment]
> We're planning on landing support for React 16.6.0 in 0.57.5. Hooks is available in the 16.7.0-alpha, and feedback from the community so far has been that upgrading to alphas of React in the stable React Native releases can be disruptive. What do people here think about a React Native 0.57.5-alpha release that includes support for 16.7.0-alpha out of the box?

I think letting people opt in to the React 16.7.0-alpha by way of an alpha release of RN is a great idea.

I've certainly struggled in the past with seeing alpha or even beta releases of React being included as part of a stable React Native release and would find it hard to justify in a more corporate environment.  However, for those people who are more adventurous, allowing access to it as an alpha release gives people the best of both worlds.

**Comment by hramos** [comment]
We've been discussing with other core contributors how to best approach this, and we're currently leaning towards focusing on landing support for React **16.6.0** stable on React Native 0.57.5 first. We're expecting 0.57.5 to be released sometime this week.

Once that happens, people who would like to experiment with React Hooks may opt to install the React 16.7.0.alpha on their project and follow the [instructions I posted earlier](https://github.com/facebook/react-native/issues/21967#issuecomment-434113687) to sync the React Native Renderer with Hooks support to their project.

As React Hooks is expected to remain experimental for a few more months, we believe this approach provides the right mix of allowing you all to opt into React Hooks, without placing additional strain on the release process for the next few months prior to React Hooks potentially reaching stable release status.

**Comment by jfrolich** [comment]
Did someone in the community that followed these steps already published a temporary package to NPM containing the alpha? Would make it easier for us to try it out by just aliasing. If not, anyone interested in it here, then I publish it myself.

**Comment by Foveluy** [comment]
for everyone who suffer from building , i have done a repo to do it, just download and copy by following 
[this  instructions ](https://github.com/facebook/react-native/issues/21967#issuecomment-434113687), you can play with hooks.

https://github.com/Foveluy/rn-temp-renderer

**Comment by IljaDaderko** [comment]
@Foveluy I got it working (I built my own react files though), few ideas

1. Are you using `const MyComponent = () =>` instead of `function MyComponent () {}` perhaps?

2. Are you calling a hook that is not directly nested in the body of component function? https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level

**Comment by Foveluy** [comment]
@IljaDaderko i tried a simple project just like
```js
function App(){
    const [xxx,setXXX] = useState(0);
    return <View>{xxx}</View>
}
```
but got the same thing showing that error.

And I built my own renderers too.

**Comment by mmazzarolo** [comment]
@Foveluy 

This `index.js` should work:
```javascript
import React, { useState } from "react";
import { AppRegistry, Text } from "react-native";
import { name as appName } from "./app.json";

const TestApp = () => {
  const [count, setCount] = useState(0);
  return <Text>{count}</Text>;
};

AppRegistry.registerComponent(appName, () => TestApp);
```

If it does not, make sure you replaced the `renderer` folder only *after* installing all the other dependencies. 
I also have to replace it each time I install a new dependency (but the error message is different from yours).

**Comment by Foveluy** [comment]
@mmazzarolo   great, thanks a lot. it is helping.

**Comment by eduplus** [comment]
So far used useState and useEffect and seems to be working just fine. Only problem is red screen while debugging, just as others have reported.

**Comment by koenpunt** [comment]
For me the instructions @hramos posted work great. But the app crashes when I try to use hot reloading. Also, when running tests it complains in certain cases about "hooks can only be used in function body", even when it is a functional component. Btw, I am using the alpha of `react-test-renderer`

**Comment by mmazzarolo** [comment]
@kodafb 
> But the app crashes when I try to use hot reloading. 

Might be related to: https://github.com/gaearon/react-hot-loader/issues/1088

**Comment by IljaDaderko** [comment]
Noticed one potential bug with useEffect and animated

```
  useEffect(
    () => {
      if (authenticated) {
        animate();
      }

      return () => {
        animation.stopAnimation();
      };
    },
    [authenticated]
  );

  function animate() {
    Animated.timing(animation, {
      toValue: 1,
      easing: easeInOutQuad,
      duration: durationLong,
      delay: 1000,
      useNativeDriver: true
    }).start();
  }
```

this is a simple animation triggered when user authenticates, works fine throughout, but when this component is unmounted and then mounted again, animation has much greater delay, expected one is `1000ms` on re-mounting (not updating) its random between `2000 - 400`

**Comment by mmazzarolo** [comment]
@IljaDaderko I'm using hooks + `Animated` as well and I'm experiencing a similar issue.  
Strangely enough, if I place a `console.warn()` in the component it works correctly.  
   
e.g.:

```javascript
export const Interlude = memo((props: Props) => {
  const { onDone } = props;
  console.warn(`Rendering interlude`); // <- Here is the console warn
  const [titleAnim] = useState(new Animated.Value(0));
  const [subtitleAnim] = useState(new Animated.Value(0));

  const animateEnter = Animated.stagger(300, [
    Animated.timing(titleAnim, {
      toValue: 1,
      duration: 500,
      useNativeDriver: true
    }),
    Animated.timing(subtitleAnim, {
      toValue: 1,
      duration: 500,
      useNativeDriver: true
    })
  ]);

  const animateExit = Animated.stagger(300, [
    Animated.timing(titleAnim, {
      toValue: 2,
      duration: 500,
      useNativeDriver: true
    }),
    Animated.timing(subtitleAnim, {
      toValue: 2,
      duration: 500,
      useNativeDriver: true
    })
  ]);

  useOnMount(() => {
    Animated.sequence([animateEnter, animateExit]).start(onDone);
  });

  const titleOpacity = titleAnim.interpolate({
    inputRange: [0, 0.5, 1, 1.5, 2],
    outputRange: [0, 0.2, 1, 0.2, 0]
  });
  const titleTranslateY = titleAnim.interpolate({
    inputRange: [0, 1],
    outputRange: [20, 0]
  });
  const subtitleOpacity = subtitleAnim.interpolate({
    inputRange: [0, 0.5, 1, 1.5, 2],
    outputRange: [0, 0.2, 1, 0.2, 0]
  });
  const subtitleTranslateY = subtitleAnim.interpolate({
    inputRange: [0, 1],
    outputRange: [20, 0]
  });
  return (
    <Animated.View style={styles.container}>
      <Animated.View
        style={{
          transform: [{ translateY: titleTranslateY }],
          opacity: titleOpacity
        }}
      >
        <Animated.Text style={styles.title}>Well done.</Animated.Text>
      </Animated.View>
      <Animated.View
        style={{
          transform: [{ translateY: subtitleTranslateY }],
          opacity: subtitleOpacity
        }}
      >
        <Animated.Text style={styles.subtitle}>+5</Animated.Text>
      </Animated.View>
    </Animated.View>
  );
});
```

I haven't tried building in release mode yet though, maybe it would solve it.

**Comment by scally** [comment]
> Let us know if the above steps work out for you.
> 

I tried following these steps and still get the `TypeError: dispatcher.useState is not a function` error. 

I can see some other folks have it working, so it must be me, but not sure what is different between my usage/env and others. 

Any chance of getting an alpha or some other prebuilt rev of this going? Now that I've seen hooks it's hard to want to keep writing things the old way.

**Comment by woodpav** [comment]
I have the same problem as @scally. I got it working before but now I can't. Perhaps recent changes in `react`?

**Comment by mmazzarolo** [comment]
@scally @woodpav I have the same issue when I install a new dependency.  
The workaround for now for me is replacing the `renderer` dir after each time I install a new dependency.

**Comment by chrisbianca** [comment]
For anybody that wants to save themselves the hassle of manually building React and remembering to update the `renderer` directory each time you install a new dependency, I have set up a Git repo with this already preconfigured.

Simply run the following and it will replace the current react-native installation with the 0.57.4 tag of RN + the hooks rendering code:

```
npm install --save https://github.com/csfrequency/react-native#0.57.4-hooks.0
```

This build will still suffer with the same Debug errors mentioned earlier on this thread.

**Comment by ericvicenti** [comment]
Thanks @chrisbianca, your build is working for me, and is the easiest way for people to try hooks in RN right now.

But I think the `useEffect` error is more serious that just a dev-mode bug. It seems to prevent the effect from working on the first few renders. This simple example should change the status bar on every tap, but it only works after one or two taps:

```js
import React, { useState, useEffect } from "react";
import { Text, View, StatusBar } from "react-native";

export default function App() {
  const [isPressed, setIsPressed] = useState(false);
  useEffect(() => {
    StatusBar.setHidden(isPressed);
  });
  return (
    <View style={{ flex: 1, justifyContent: "center" }}>
      <Text
        onPress={() => {
          setIsPressed(!isPressed);
        }}
      >
        {isPressed ? "Wohoo Hooks" : "Press Me"}
      </Text>
    </View>
  );
}
```

**Comment by chrisbianca** [comment]
I've just pushed an update which removes the annoying Debug errors.  You can install from a new tag:

```
npm install --save https://github.com/csfrequency/react-native#0.57.4-hooks.1
```

@ericvicenti This doesn't address the issue you've flagged above, but should at least make debugging it easier.

**Comment by IljaDaderko** [comment]
@chrisbianca Neat! What was the cause of that?

**Comment by woodpav** [comment]
For anyone who would like to try hooks in production replace all instances of `enableHooks = false` with `enableHooks = true` in the cloned `react` repo. The relevant files are in `react/packages/shared`.

The `yarn build` command uses these flags to create `ReactNativeRenderer-prod.js` so you have to `enableHooks` before copying those files to your projects `node_modules`.

**Comment by jas99** [comment]
Hey guys
You can also use my starter project.
https://github.com/jas99/starter-ts-react-native-hook
It uses monkey-patching and also have typescript setup.

Thx

**Comment by brentvatne** [comment]
I took @chrisbianca's branch and applied one more patch to it in order to get it running in Expo, you can try the project out here if you like: https://github.com/expo/sdk-31-with-hooks-experimental

**Comment by brunolemos** [comment]
There were new releases today, `react-native@0.57.5` and `react@16.7.0-alpha.2`. I know you guys will probably update your builds soon but I did my own as well: 

```bash
yarn add react@next react-native@"npm:@brunolemos/react-native"
yarn add --dev @types/react @types/react-native
```

I published it on npm instead of git, the installation seems faster this way.

[Tweet](https://twitter.com/brunolemos/status/1062476577160785922)

**Comment by brunolemos** [comment]
I tested @ericvicenti's code. There seems to be two issues:

- The effect call seems to be one render late. It seems to happen only when remote debug is enabled.
- When remote debug is disabled, it works, but the effect takes a couple seconds to happen.

**Comment by jfrolich** [comment]
@IljaDaderko: `react-spring` does already support hooks, and it even offers a bit more functionality, so might worth having a look at that.

**Comment by IljaDaderko** [comment]
> @IljaDaderko: `react-spring` does already support hooks, and it even offers a bit more functionality, so might worth having a look at that.
> 
> Thanks for releasing the hooks version of react-native, I didn't have time to work on this.

Indeed, but I'm trying to utilise rn API as much as I can at the moment :)

**Comment by brunolemos** [comment]
@mmazzarolo did you also face it or just referring to the gif?

**Comment by mmazzarolo** [comment]
@brunolemos yep, facing it from day one.  
Strangely enough, adding a `console.warn()` in the component that uses `useEffect` fixes the issue.

**Comment by IljaDaderko** [comment]
Noticed another `useEffect` related issue. Code below works as expected when `delay` property is not passed to animated i.e. animation finishes and `logoAnimationLogic` is called. If I add `delay` (no matter what value), then `logoAnimationLogic` is not called after animation has finished for roughly 2 - 3 seconds (again this doesn't seem to be related to delay value).
 
```js
function logoAnimationLogic({ finished }) {
    if (finished) {
      setVisible();
    }
  }

 useEffect(
    () => {
      if (authenticated) {
        Animated.timing(animatedValue, {
          delay: 100,
          duration: 600,
        }).start(logoAnimationLogic);
      }
    },
    [authenticated]
  );
```

**Comment by jfrolich** [comment]
> @jfrolich unfortunately it looks like react-spring [doesn't support `useNativeDriver` yet](https://github.com/drcmda/react-spring/issues/299)

Thanks I was considering to use react-spring in react native, but indeed if it doesn't support useNativeDriver it's not a great option.

**Comment by jfrolich** [comment]
Does anyone have the problem that component trees with hooks don't hot reload?

**Comment by mmazzarolo** [comment]
> Does anyone have the problem that component trees with hooks don't hot reload?

I do. I think it might still be related to the `useEffect` issue discussed above.

**Comment by IljaDaderko** [comment]
Not sure if anyone tried it yet but issues outlined here
https://github.com/facebook/react-native/issues/21967#issuecomment-438483674

Also happen in release builds and on physical devices, so its likely not linked to dev mode or simulator.

**Comment by jas99** [comment]
@hramos Curious about timetable on this, any updates or this is on back burner awaiting react?

Thx

**Comment by woodpav** [comment]
@slorber I believe [this line](https://github.com/slorber/react-native-animation-hooks/blob/da3ed835b22a1a485c15b3fff7f8aeeb633a8cd5/src/AnimationHooks.ts#L11) should be `useState`.

Try `const [animatedValue] = useState(new Animated.Value(initialValue));`

**Comment by IljaDaderko** [comment]
> @slorber I believe [this line](https://github.com/slorber/react-native-animation-hooks/blob/da3ed835b22a1a485c15b3fff7f8aeeb633a8cd5/src/AnimationHooks.ts#L11) should be `useState`.
> 
> Try `const [animatedValue] = useState(new Animated.Value(initialValue));`

It can also be ref, you don't necessarily need to store animated value in state, but you can.

**Comment by mmazzarolo** [comment]
@woodpav I think using a `ref` is ok

**Comment by bvaughn** [comment]
Based on @slorber's comment above, I've created this local test:
https://gist.github.com/bvaughn/b590a6bfe9ae06fd100dea2a77ab3775

But I'm not seeing any problems with `useEffect`. Anything I'm overlooking perhaps?

**Comment by IljaDaderko** [comment]
@bvaughn I haven't tried example posted by @slorber but @mmazzarolo posted yield same result as his gif previews.
https://github.com/facebook/react-native/issues/21967#issuecomment-439641251

**Comment by bvaughn** [comment]
Hm...the code in @mmazzarolo's comment is a lot larger (and TypeScript based) so I was hoping to be able to repro `useEffect` issue with something smaller.

**Comment by bvaughn** [comment]
I wonder if anyone else in this thread that's seeing the problem can reproduce it with [the small Gist](https://gist.github.com/bvaughn/b590a6bfe9ae06fd100dea2a77ab3775) I shared?

**Comment by brunolemos** [comment]
@bvaughn just tested your gist and I saw the exact same bugs that I commented [here](https://github.com/facebook/react-native/issues/21967#issuecomment-438483674)

If you don't, please share your build with us.

`package.json`
```json
"react": "16.7.0-alpha.2",
"react-native": "npm:@brunolemos/react-native",
```

**Comment by bvaughn** [comment]
Thanks for confirming, @brunolemos.

So far, I have been trying to reproduce with the version of React Native that's running within Facebook, which was built from [React commit `3ff2c7c`](https://github.com/facebook/react/commit/3ff2c7c) about 3 weeks ago.

**Comment by slorber** [comment]
@bvaughn maybe my example is a bit too subtile because the animation seems to work, but actually it works in the wrong way: when the toggle is true, the items are in the falsy positions, and vice versa.

If you run the Expo app of the repo, and add logs to the `useAnimation` entry and `useEffect` method, you will see that when `useAnimation` is called, the `useEffect` method is not called just after while it should. The `useEffect` is only called once we press again on the toggle, like if the effects didn't flush on update but rather flushing on next render/interaction or something

**Comment by bvaughn** [comment]
@slorber That's interesting. Changing [my gist](https://gist.github.com/bvaughn/b590a6bfe9ae06fd100dea2a77ab3775) to also show the current state of the bool should show that then, no?
```js
<Text onTouchEnd={toggle}>Bool is {bool ? 'on' : 'off'}</Text>
```
But it doesn't seem to. "Bool is on" always matches the fully opaque state.

**Comment by bvaughn** [comment]
@gaearon I also tried reproducing with the OSS build of React Native (which I manually edited to set `enableHooks = true`) but saw the same results.

**Comment by gaearon** [comment]
Yes, I meant specifically the GH setup, which AFAIK has its own lockfile etc.

**Comment by bvaughn** [comment]
I am able to reproduce the bug with my Gist running in an Expo app. 👍

Going to try to dig into the differences now.

**Comment by bvaughn** [comment]
I also seem to be able to reproduce the bug when connected to the remote debugger (Chrome) although I had to manually edit the `scheduler` package to patch a `postMessage` bug.

**Comment by bvaughn** [comment]
What I'm seeing in my test harness seems to be that the state update caused by calling `toggle` flushes the previous update's `useEffect` call, so the animate function is always running one version behind, like @slorber said.

**Comment by bvaughn** [comment]
Ah, so this is interesting. When I more properly patch `scheduler` (aka just replace it with [the canary release](https://unpkg.com/scheduler@canary/cjs/scheduler.development.js)) then the bug goes away when connected to the Chrome debugger, but still happens when I'm not connected.

However in both cases, `scheduler` seems to be using the `MessageChannel` approach so I'm not sure yet why I'd see different behavior.

**Comment by bvaughn** [comment]
I think I've found the cause. [The `setTimeout` branch of `scheduler`](https://github.com/facebook/react/blob/c2a2d8a539bf02e40c43d36adc2826e228f30955/packages/scheduler/src/Scheduler.js#L463-L495) is scheduling callbacks on a 5000ms delay. This 5000ms is _supposed to be the maximum expiration time_, but in this fork it ends up being the _minimum callback time_ (unless another state update forces us to sync-flush pending callbacks). So the result is that React doesn't call `useEffect` until after a few seconds in that branch.

I assume we (React team) haven't noticed this because we're typically using the `postMessage` implementation, but on iOS with JSC– React Native ends up using the `setTimeout` fork.

Unless I'm misreading things, that's the problems.

**Comment by jas99** [comment]
Can anybody confirm useaffect bug only affects ios?

**Comment by ericvicenti** [comment]
@jas99, the `useEffect` bug(s) are reported on Android as well by @slorber here: https://github.com/slorber

**Comment by bvaughn** [comment]
This issue impacts JavaScriptCore, which is what React Native uses on Android as well.

**Comment by bvaughn** [comment]
Fix has landed via facebook/react/pull/14358 and will go out with the next release of `scheduler`.

If it's necessary, we can also do a bugfix release to the current `scheduler` but I think the number of people this affects is probably very low so I'm not sure how important it is to do that patch release?

**Comment by brunolemos** [comment]
@bvaughn awesome 🎉🎉

I know some people that postponed playing with it on react native because of the bug, so maybe make the patch fix? Which should increase the number of testers.

**Comment by kevinwolfcr** [comment]
For the ones having problems with `useEffect`, I switched to `useLayoutEffect` (while facebook/react#14358) gets merged, and it works.

**Comment by bvaughn** [comment]
Yup, `useLayoutEffect` is synchronous and so it isn't affected by the timer bug.

**Comment by brunolemos** [comment]
Thanks @bvaughn! 🎉 

It's perfect now. 

For anyone else, I made a new build based on `react-native@0.57.7`: 

```bash
yarn add react@next react-native@"npm:@brunolemos/react-native"
yarn add --dev @types/react @types/react-native
```

Also add this to your `package.json`:

```json
  "resolutions": {
    "scheduler": "0.12.0-alpha.3"
  },
```

<img height="300" src="https://user-images.githubusercontent.com/619186/49399688-1f8f3980-f729-11e8-8f61-7cace37ca1d2.gif" />

**Comment by bvaughn** [comment]
Niiiiice!

**Comment by sunnylqm** [comment]
@brunolemos Thanks for your build, but I think the resolutions part is unnecessary. `scheduler@^0.12.0-alpha` would be resolved to the latest alpha version.

**Comment by kevinwolfcr** [comment]
Awesome, @bvaughn! Do you know if it works with Expo?

**Comment by bvaughn** [comment]
> Awesome, @bvaughn! Do you know if it works with Expo?

I don't have a lot of Expo experience (in fact, my first time using it was looking into this bug, trying to run the repro) but... why wouldn't it? Any particular reason?

**Comment by kevinwolfcr** [comment]
@bvaughn that's because as Expo runs on its own app, sometimes they do modifications to the `react-native` package, that's because when you are on an expo project, the dependency `react-native` its resolved to something from https://github.com/expo/react-native/releases

---

@brunolemos I am using

```
"react-native": "https://github.com/expo/react-native/archive/sdk-31-with-hooks-experimental.tar.gz",
```

on my package.json. Everything works fine, except trying to build the app, which is a huge blocker :(

**Comment by josemigallas** [comment]
@brunolemos I installed your package but `useState` in undefined.. I get this warning when `yarn install`ing:
>warning Resolution field "scheduler@0.12.0-alpha.3" is incompatible with requested version "scheduler@^0.11.0"

However I added the 'resolutions' prop in my package.json. Did you do any other step? 😐
This is my `package.json`:
```
{
  "name": "RealmsCounter",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "start": "node node_modules/react-native/local-cli/cli.js start",
    "test": "jest"
  },
  "dependencies": {
    "@brunolemos/react-native": "npm:@brunolemos/react-native",
    "react": "^16.7.0-alpha.2",
    "react-native": "npm:@brunolemos/react-native"
  },
  "devDependencies": {
    "babel-jest": "23.6.0",
    "jest": "23.6.0",
    "metro-react-native-babel-preset": "0.50.0",
    "react-test-renderer": "16.6.1"
  },
  "jest": {
    "preset": "react-native"
  },
  "resolutions": {
    "scheduler": "0.12.0-alpha.3"
  }
}
```
Also I'm using
```
yarn 1.12.3
react-native-cli: 2.0.1
react-native: 0.57.7-hooks.1
```

**Comment by bySabi** [comment]
@josemigallas I think `scheduler` was update on latest react-native RC  0.58.0-rc.1

test it again with:
`yarn add react-native@next`

**Comment by alk831** [comment]
![hooks error](https://user-images.githubusercontent.com/619186/47597428-fed30780-d964-11e8-84a9-b60c874063ae.png)

How do I solve this issue? Im using `0.57.8` version of React Native, `16.7.0-alpha.0` v of React and React Test Renderer and I don't know what causes this error.

**Comment by woodpav** [comment]
@alk831 Make sure you replace all instances of `enableHooks = false` with `enableHooks = true` in `react/packages/shared` before running `yarn build -- --type=RN_OSS`

**Comment by brunolemos** [comment]
I've been using Hooks in production with zero problems so far! 
Using [this method](https://github.com/facebook/react-native/issues/21967#issuecomment-443858969) I posted earlier.

You can check the source code [here](https://github.com/devhubapp/devhub) and the iOS app [here](https://itunes.apple.com/br/app/devhub-for-github/id1191864199?l=en&mt=8). Haven't published the Android app yet.

**Comment by rmilejcz** [comment]
@brunolemos  thanks for that, it works great even `useEffect` is working. Do you plan on keeping that repo up to date with new `rn` releases? Thanks again!

**Comment by kevinwolfcr** [comment]
@MrLoh actually it is a release of expo/react-native: https://github.com/expo/react-native/releases.

What version of RN are you using on your project?

**Comment by MrLoh** [comment]
@kevinwolfcr, I just forked the code from that repo. Check the package.json there.

**Comment by sarmad1995** [comment]
@brunolemos I followed your steps and it's working, but now when i use npm install --save somelibrary,
it seems like my npm always fails..any reason?

**Comment by jas99** [comment]
@sarmad1995 https://github.com/ds300/patch-package

**Comment by jas99** [comment]
@sarmad1995 after npm install look back in node_modules/react-native; ur changes r gone, use above package to monkey patch

**Comment by bntzio** [comment]
Mine is working perfectly, without expo.

**Comment by davibe** [comment]
Can someone explain how can i do this ? 

I have downloaded an expo project and then i have tried doing this.

```
yarn add react@next react-native@"npm:@brunolemos/react-native"
yarn add --dev @types/react @types/react-native
```
added to package.json
```
"resolutions": {
  "scheduler": "0.12.0-alpha.3"
},
```

however this is the error i get when i try to run it with `expo start`
```
Unable to resolve "react-native/Libraries/Components/View/ViewStylePropTypes" from "node_modules/react-native-reanimated/src/createAnimatedComponent.js"
ERROR
18:38
Building JavaScript bundle: error
```

**Comment by bvaughn** [comment]
The error message you pasted seems to suggest that Yarn had a problem installing @brunolemos fork of react-native, because the path the error message references (`react-native/Libraries/Components/View/ViewStylePropTypes`) [seems valid](https://unpkg.com/react-native/Libraries/Components/View/ViewStylePropTypes.js).

**Comment by brunolemos** [comment]
@davibe for expo use this instead of my fork https://github.com/expo/sdk-31-with-hooks-experimental

**Comment by JPabloObando** [comment]
> Hooks es un gran cambio para la sintaxis React / RN. ¿Una gran cantidad de documentos necesitarán una actualización también?

It is optional to implement it because imagine that Hooks is a functional component with access to the state, to the life cycle, to the context. It is a "class with a wrapped in a functional component". The classes are not going to disappear or the functional components, now we have an easier way to use what the class provided us and to reuse logic.

**Comment by colmartofus** [comment]
I made a stackoverflow post with a small guide here:

https://stackoverflow.com/questions/53960785/typeerror-dispatcher-usestate-is-not-a-function-for-react-native

**Comment by chinds185a** [comment]
Wonder if anyone can help, I have followed @brunolemos guide on getting up and running with hooks on react-native, but as soon as I start to use hooks in my project I run into this error in the android simulator;

![image](https://user-images.githubusercontent.com/11033500/50560528-d0253480-0cf9-11e9-9ad0-57327b9c0f5e.png)

**Comment by brunolemos** [comment]
@chinds185a I'm assuming you already did this: https://github.com/facebook/react-native/issues/21967#issuecomment-449615358

And followed the upgrade guide of each react native version after the one you were using.

Upgrade your metro-react-native-babel-preset version. Change your .babelrc to this https://github.com/devhubapp/devhub/blob/master/.babelrc.

It seems you are using custom jsc. Remove the polyfills, make sure you followed the steps correctly: https://github.com/react-native-community/jsc-android-buildscripts/blob/master/README.md#how-to-use-it-with-my-react-native-app

**Comment by chinds185a** [comment]
Thanks for the suggestions @brunolemos will try these later. 

I have followed the upgrade path as this is a new project and I started with the latest versions of react and react-native.

**Comment by chinds185a** [comment]
@brunolemos Thanks for the suggestions however I am still having the same issues, and it only appears when trying the `useContext` hook :/

I have;

- Removed the polyfills
- Locked react version too `16.7.0-alpha.2`
- Updated `metro-react-native-babel-preset` to `^0.51.1`
- Using the latest JSC for android and followed the setup/build instructions
- I am running babel 7
- Have react-native installed from `npm:@brunolemos/react-native`

**Comment by rmilejcz** [comment]
@chinds185a is your package.json the same as you previously posted? You need to remove the caret from `^16.7.0-alpha.2`, it should be `16.7.0-alpha.2` instead since `16.7.0` is officially released. Might also be a good idea to delete `node_modules` and then reinstall

**Comment by gazedash** [comment]
Any estimate about the date when hooks will be ready to use with RN in production? I mean, using stable version, not alphas and betas

**Comment by pvinis** [comment]
I think it should be this coming 0.58 right?

**Comment by rmilejcz** [comment]
@pvinis I doubt it.

@gazedash 0.58 will probably ship with react 16.7 as a dependency, which does not include support for hooks unfortunately. Ideally, 16.8 *will* include support for hooks, and hopefully we can see that by react-native 0.60. For now we'll just have to accept that hooks are an experimental feature in react  so using them is gonna take a bit of legwork

**Comment by bvaughn** [comment]
The above replies are correct!

We (React core team) haven't announced the date when hooks will be released in a stable React version– (other than saying "soon")– and this is a prerequisite for supporting hooks in React Native.

I will say that as soon as hooks are released in React stable, I'll be working with the React Native team to unblock anything needed to support hooks there. I'm already working on upgrading ancillary packages (like the [standalone `react-devtools`](https://www.npmjs.com/package/react-devtools)) in preparation.

**Comment by aleclarson** [comment]
@bvaughn Are there large changes required for React Native to support hooks, or is it pretty basic stuff? I'm using [ptmt/react-native-macos](https://github.com/ptmt/react-native-macos), and I'm considering using React hooks in my codebase, even if that means making some changes to native code.

**Comment by bvaughn** [comment]
There might some unknowns but I don't think it should be large, no.

**Comment by hramos** [comment]
For those asking for a timeline, support for Hooks in RN should land no earlier than Release 0.60. This is based on the fact the 0.59 RC is due to be cut soon, with React 16.6.3 as a dependency most likely.

**Comment by fbartho** [comment]
Oh no‼️ ❤️ hooks, is there anywhere where we can beg to get hooks into 59? @hramos

**Comment by hramos** [comment]
The 0.59 release is being coordinated in https://github.com/react-native-community/react-native-releases/issues/79, and that's the best place to discuss if you feel strongly that this cannot wait for 0.60.

**Comment by kelset** [comment]
If you want a more in depth explanation of what it takes for Hooks to land: https://github.com/react-native-community/react-native-releases/issues/79#issuecomment-457735214

(locking to keep this as last link until updates)

**Comment by hramos** [comment]
React 16.8 with Hooks has been released: https://reactjs.org/blog/2019/02/06/react-v16.8.0.html

We are working on updating React Native to use React 16.8 in time for the React Native 0.59 release.

**Comment by hramos** [comment]
👀 https://github.com/facebook/react-native/commit/ccefc700d0120539eba73747d1d6b65effb0645d

**Comment by hramos** [comment]
The 0.59 release has been promoted to stable. Read more in the React Native blog: http://facebook.github.io/react-native/blog/2019/03/12/releasing-react-native-059

Now that Hooks are generally available, I'll go ahead and unpin this thread.
